cmd/influxd/restore/command.go

```diff
@@ -0,0 +1,239 @@
+	Args: cobra.ExactArgs(0),
+	RunE: restoreE,
+}
+
+var flags struct {
+	boltPath   string
+	enginePath string
+	backupPath string
+	rebuildTSI bool
+}
+
+func init() {
+	dir, err := fs.InfluxDir()
+	if err != nil {
+		panic(fmt.Errorf("failed to determine influx directory: %v", err))
+	}
+
+	Command.Flags().SortFlags = false
+
+	pfs := Command.PersistentFlags()
+	pfs.SortFlags = false
+
+	opts := []cli.Opt{
+		{
+			DestP:   &flags.boltPath,
+			Flag:    "bolt-path",
+			Default: filepath.Join(dir, bolt.DefaultFilename),
+			Desc:    "path to target boltdb database",
+		},
+		{
+			DestP:   &flags.enginePath,
+			Flag:    "engine-path",
+			Default: filepath.Join(dir, "engine"),
+			Desc:    "path to target persistent engine files",
+		},
+		{
+			DestP:   &flags.backupPath,
+			Flag:    "backup-path",
+			Default: "",
+			Desc:    "path to backup files",
+		},
+		{
+			DestP:   &flags.rebuildTSI,
+			Flag:    "rebuild-index",
+			Default: true,
+			Desc:    "if true, rebuild the TSI index and series file based on the given engine path (equivalent to influxd inspect build-tsi)",
+		},
+	}
+
+	cli.BindOptions(Command, opts)
+}
+
+func restoreE(cmd *cobra.Command, args []string) error {
+	if flags.backupPath == "" {
+		return fmt.Errorf("no backup path given")
+	}
+
+	if err := moveBolt(); err != nil {
+		return fmt.Errorf("failed to move existing bolt file: %v", err)
+	}
+
+	if err := moveEngine(); err != nil {
+		return fmt.Errorf("failed to move existing engine data: %v", err)
+	}
+
+	if err := restoreBolt(); err != nil {
+		return fmt.Errorf("failed to restore bolt file: %v", err)
+	}
+
+	if err := restoreEngine(); err != nil {
+		return fmt.Errorf("failed to restore all TSM files: %v", err)
+	}
+
+	if flags.rebuildTSI {
+		sFilePath := filepath.Join(flags.enginePath, storage.DefaultSeriesFileDirectoryName)
+		indexPath := filepath.Join(flags.enginePath, storage.DefaultIndexDirectoryName)
+
+		rebuild := inspect.NewBuildTSICommand()
+		rebuild.SetArgs([]string{"--sfile-path", sFilePath, "--tsi-path", indexPath})
+		rebuild.Execute()
+	}
+
+	if err := removeTmpBolt(); err != nil {
+		return fmt.Errorf("restore completed, but failed to cleanup temporary bolt file: %v", err)
+	}
+
+	if err := removeTmpEngine(); err != nil {
+		return fmt.Errorf("restore completed, but failed to cleanup temporary engine data: %v", err)
+	}
+
+	return nil
+}
+
+func moveBolt() error {
+	if _, err := os.Stat(flags.boltPath); os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if err := removeTmpBolt(); err != nil {
+		return err
+	}
+
+	return os.Rename(flags.boltPath, flags.boltPath+".tmp")
+}
+
+func moveEngine() error {
+	if _, err := os.Stat(flags.enginePath); os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if err := removeTmpEngine(); err != nil {
+		return err
+	}
+
+	if err := os.Rename(flags.enginePath, tmpEnginePath()); err != nil {
+		return err
+	}
+
+	return os.Mkdir(flags.enginePath, 0777)
```

Comment:
I think this should be `os.MkdirAll(flags.enginePath, 0777)` to ensure the parents are made too. Since we check to see if `/var/lib/path/to/data/engine` exists in L145, we should make sure that this is fully created. Alternatively if we're just changing the perms we should use `Chmod`?

https://golang.org/pkg/os/#Chmod

Reply:
I would rather not worry about chmod'ing an existing dir here. I agree that `MkdirAll` makes sense.